### ISSUE TITLE
BSOD when combining WinDivert with Avast (part II)

### ISSUE THREAD

**Issue by basil00** [basil00/Divert#110]
*Note: This is not affect WinDivert 1.3.*

Version 1.4 of WinDivert has been delayed because the Avast BSOD issue has arisen again.  This was strange, since the original code that "caused" the previous issue (#90) had been fixed.

After some investigation, it seems that the issue relates to a call to `NdisRetreatNetBufferDataStart` in the [`windivert_worker`](https://github.com/basil00/Divert/blob/af5cd698837a679c85198de678be28f7225538aa/sys/windivert.c#L2623) function.  If this call is removed, WinDivert and Avast run together smoothly without issue, although WinDivert's functionality is broken since it is reading the packet buffer from the wrong offset.  Also, if WinDivert is restricted to outbound packets only, then `NdisRetreatNetBufferDataStart` is never called, and again there is no issue with WinDivert and Avast.  Furthermore, this would partially explain (#90), since the old driver was not resetting the data start when the classify function exits.

Now the question is: is it valid to call `NdisRetreatNetBufferDataStart` on a packet blocked & absorbed by a callout driver?  The current out-of-band design of WinDivert1.4 was based on Microsoft's WFPSampler drivers, which do indeed call `NdisRetreatNetBufferDataStart` (e.g., see [here](https://github.com/Microsoft/Windows-driver-samples/blob/2cbdf8a4423e8c1c72e049740d940b627e2ad29f/network/trans/WFPSampler/sys/ClassifyFunctions_AdvancedPacketInjectionCallouts.cpp#L1462)) out-of-band for blocked & absorbed packets (for the same reason as WinDivert needs to).  Furthermore, it seems that the only thing the WFPSampler does to the original packet is simply [reference it](https://github.com/Microsoft/Windows-driver-samples/blob/aa6e0b36eb932099fa4eb950a6f5e289a23b6d6e/network/trans/WFPSampler/syslib/HelperFunctions_ClassifyData.cpp#L344), which is exactly what WinDivert does too.

So unless I am missing something, the design at least should be sound.  The point of block & absorb was to hand ownership of the packet over to the driver (WinDivert) for further out-of-band processing.  All other processing on the packet should be stopped.

At this point I attempted to compile the WFPSampler drivers to test if they are compatible with Avast.  After a long process, I was able to build the drivers and executables, but the `WFPSampler.exe` file refuses to run with a generic `The system cannot execute the specified program` error.  I am not sure what has gone wrong.

So this is where I am currently stuck.  If anyone knows how to compile and build the WFPSampler drivers/programs then please let me know.  Unfortunately it seems that Microsoft do not provide pre-compiled versions.

**Comment by Areithus** [comment]
Did you run the WFPSamplerInstall.cmd script? It should be executed before you run the executable.

**Comment by kolaci** [comment]
Is possible to download memory dump file for this issue?

**Comment by basil00** [comment]
@Areithus Yes I did run the install script.  After prompted for reboot (&did so), still no luck.  I guess that I must have messed up my VS2015 + WDK install.

@kolaci Thanks for taking a look.  I can send you links to crash dumps and/or the (signed) WinDivert1.4 drivers that exhibit the problem.  Can you contact me by email (my email address is at the bottom of [this page](https://reqrypt.org/windivert.html))?

**Comment by polachz** [comment]
@basil00 @kolaci From my previous experience with WFP, i can say that reference just "protect" data against deletion. But not grants data consistency between calls.  Then referenced packet can be changed between point of referencing and next OOB processing. Often the referenced packet changes it's length etc. It happened to me on STREAM_LAYER, but i think that this behavior is generic.
If you have to grant packet in same state as in time of first processing (referencing) you have to provide clone by my experience. Or, you can delay to move Retreat/Advance mechanism to OOB processing.
Additionally, do you check the FWPS_RIGHT_ACTION_WRITE flag in callout? From WFP examples, if the flag is not set, they do not provide any action on the packet. It's jist notification that other driver process the packet.

**Comment by basil00** [comment]
> Then referenced packet can be changed between point of referencing and next OOB processing.

I think this is not supposed to happen (unless the other drivers are buggy).

WinDivert 1.4 is an "Out-of-band Modification Callout" as per [Microsoft's definition](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/types-of-callouts), so the procedure is to `FWP_ACTION_BLOCK` and `FWPS_CLASSIFY_OUT_FLAG_ABSORB` the packet, and take a reference using `FwpsReferenceNetBufferList`.  After that it is "ready to be modified outside classifyFn".  WinDivert also clears the `FWPS_RIGHT_ACTION_WRITE` bit ([here](https://github.com/basil00/Divert/blob/af5cd698837a679c85198de678be28f7225538aa/sys/windivert.c#L2584)) meaning that other callouts should not inspect nor modify the packet.

The only difference with Microsoft's documentation is that WinDivert does not use `FWP_ACTION_CALLOUT_TERMINATING` since the classify function can return `FWP_ACTION_CONTINUE`.  I tried changing this but it does not fix the BSOD, so that is not the issue.

> Or, maybe (not tested) you can delay to move Retreat/Advance mechanism to OOB processing.

This is what WinDivert is doing.  As noted in my original post, it seems this is a valid thing to do since even the Microsoft samples do something similar.  It seems that the BSOD is the result of the packet being used/modified/inspected by someone else despite WinDivert supposedly having complete ownership of it.  This could also be caused by WinDivert not correctly taking ownership of the packet.  However, staring at the code for a long time I cannot see where it goes wrong.

> Additionally, do you check the FWPS_RIGHT_ACTION_WRITE flag in callout?

WinDivert checks the flag [here](https://github.com/basil00/Divert/blob/af5cd698837a679c85198de678be28f7225538aa/sys/windivert.c#L2370), and takes no action as you say.  Perhaps this is something @kolaci can check on his end?  E.g. is the `FWPS_RIGHT_ACTION_WRITE` cleared, and if so, does your driver take no action?

**Comment by polachz** [comment]
@basil00 

I an just say, that reference brings BSODs to my code, when I move to clone, it gone. I used Microsoft example as base of my code too. It doesn't grant that MS example i correct in all way. I got indirect confirmation that examples can be buggy too from MS Staff on the Microsoft Plug Fest.

**Comment by basil00** [comment]
Cloning will probably "work" because advancing/retreating no longer alters the state of the original `NET_BUFFER`.  It is not as efficient as simply taking a reference, but I may fall back to this approach if nothing else works.

That said, I still want to find the root cause of the problem.  It is apparent that the "fix" from version 1.3 didn't really do this.  If the same bug exists in the samples or Windows itself then I doubt Microsoft will just find this acceptable.

> From brief observation, you "redirect" the packet from net card stack to loopback, right? I think that it means big change and clone is better way to do that.

No, the packet is queued and processed in a work item.  This is *out-of-band* processing and is a standard approach for implementing WFP callout drivers, at least according to the documentation and samples.

**Comment by basil00** [comment]
Just a quick update: development is on hold until the new year.  If there is still no resolution to the bug, I'll switch to using cloned packets for the 1.4 release as per @polachz's suggestion.

**Comment by basil00** [comment]
After further analysis it seems that the hypothesis in the original post is correct, i.e. for whatever reason, the Avast driver is modifying packets (via `NdisRetreatNetBufferDataStart`) that were blocked & absorbed by WinDivert.  This leads to a race: where the WinDivert out-of-band worker thread and the Avast driver are both trying to advance/retreat the same packet at the same time, and this is very likely the underlying cause of the BSOD.

Whatever the reason for the conflict, it seems like a good idea to switch WinDivert back to using inline packet processing & deep copying.  Firstly, this completely removes the possibility of such bugs.  Secondly, it helps simplify the driver somewhat.  Finally, it works better for sniff mode since we can just "continue" then original packet rather than reinject a copy out-of-band.

The downside is that WinDivert is now using deep copying again, like WinDivert 1.3. That said, there did not appear to be any significant performance advantage from version 1.3 and out-of-band processing.

**Comment by polachz** [comment]
Hi @basil00.

I think taht i have found the problem after your description. It' design problem and it explain why the Cone is necessary instead of the reference.
When the First Callout is called, you have guarantied  by MS specification, that packet is in expected state. MS require by specification, that you have to restore packet to the same state, before you leave the callout, even you will absorb it or not. Inside the callout call, you have permission to modify headers by retreat/Advance etc, but you have to restore this before leave.
But when you leave the callout, you loose control over the original NBL, because other drivers can do same things inside their callouts. MS grants that just one callout is called over one NBL. No concurrency is allowed here. But MS can't grant this for OOB processing, because hasn't control over OOB threads.
If you just reference the packet, you just block it against delete, but not protect it against other drivers processing. And then you can get race during your OOB thread call. So, if you need to do ANY changes in NBL in OOB thread, you have to do over your clone, not original NBL.
I think that it explains why the reference is not possible for OOB with any modifications.
Best regards,
Zdenek

**Comment by basil00** [comment]
I am not sure since several of Microsoft's own sample programs, including [Out-of-band Packet Modification from Incoming Datagram Data Layers](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/packet-modification-examples), merely reference the NBL and retreat/advance out-of-band, which is precisely what caused the problem with WinDivert+Avast.

Was it ever clarified that the Avast driver is respecting the `FWPS_RIGHT_ACTION_WRITE` flag?  WinDivert clears the flag, as does every Microsoft OOB-processing sample sample, so the flag should also be cleared once the packet is indicated to the Avast driver.  Furthermore, no Microsoft sample modifies the packet when the flag is cleared, so there may be an implied rule that drivers are not supposed to.  This would also resolves the race.

**Comment by polachz** [comment]
Forget on MS example ,please. It's buggy sometimes, we know this.

**Comment by kolaci** [comment]
Hi Basil,
According to documentation https://msdn.microsoft.com/en-us/library/windows/hardware/ff551229(v=vs.85).aspx, flag FWPS_RIGHT_ACTION_WRITE controls access to member FWPS_CLASSIFY_OUT::actionType only and Avast driver works exactly this way. So I think, you cannot use the FWPS_RIGHT_ACTION_WRITE for protection whole NBL.

**Comment by basil00** [comment]
@polachz No I do not think it is likely that the Microsoft samples are buggy.  But if you believe so, then please submit a bug report and they will definitely want to fix them.  Remember that other WFP devs will be basing their drivers off the same code, and this will cause problems for your users.

@kolaci The way I see it: the Microsoft samples consistently use the `FWPS_RIGHT_ACTION_WRITE` bit to protect the entire NBL, and using it in this way avoids race conditions, so this is *probably* the intended use case.  Although the link does not mention it one way or the other, it also doesn't exclude this possibility either.  Other [documentation](https://msdn.microsoft.com/en-us/library/windows/hardware/ff544890(v=vs.85).aspx) and various samples suggest `FWPS_RIGHT_ACTION_WRITE` can be used to protect referenced NBLs.

If your interpretation is wrong then it means there is a very serious bug in the Avast driver, and is the underlying cause of the crashes #90 and #110.  If your interpretation is correct, then it means the Microsoft samples or perhaps the Windows kernel itself contains bugs, so they'd definitely want to hear about it.

**Comment by kolaci** [comment]
We are anti-virus SW, we need to see every packets including these which were blocked by another SW (these packets should have FWPS_RIGHT_ACTION_WRITE set to FALSE too). We need to track and report every found virus signature to our anti-virus engine independently on third-party SW. This is why I cannot ignore packets with cleared FWPS_RIGHT_ACTION_WRITE bit.

**Comment by basil00** [comment]
I understand why you want to do this, however that is not relevant if it is not allowed by the API.  This is why I suggested getting clarification from Microsoft.

Either way, this bug affects your driver and your users, so it is up to you to fix it.  WinDivert ignores packets with the `FWPS_RIGHT_ACTION_WRITE` bit cleared so is not affected by this problem.

**Comment by TechnikEmpire** [comment]
100 percent agree with basil and I offer the same advice. Right now the advice to a community of about 25K people is to switch antivirus software so it doesn't break their other software that uses WinDivert. I'm sure that's gotta be more valuable than sending a few dropped packets to your servers for analysis.

**Comment by polachz** [comment]
@basil00 Can you please give me a link, where you find in MS doc that the flag FWPS_RIGHT_ACTION_WRITE Means that you have to leave the packet untouched?? I just found this:

FWPS_RIGHT_ACTION_WRITE

    If this flag is set, a callout driver can write to the actionType member of this structure. If this flag is not set, a callout driver can write only to the actionType member of this structure if it is vetoing an FWP_ACTION_PERMIT action that was previously returned by a higher weight filter in the filter engine.

here:
https://msdn.microsoft.com/en-us/library/windows/hardware/ff551229(v=vs.85).aspx

So, we can't modify FWPS_CLASSIFY_OUT0 but nothing else (and it's what we exactly does).

Then, your idea that these packet can't be analysed, is completely wrong By my opinion. Why the MS sent it to callout if you can't touch it anyway? 

I have written several WFP drivers, and my experience is that Clone is necessary for OOB processing. We solved interop problems with other products by this way. It's recommendation which I got from MS staff on the PlugFest events.
Additionally, if you take look on Clone operation, it's not a deep copy of data. It just allocate new NBL structure and attach it to existing MDL with network data. With right NBL pool it's very efficient.

**Comment by kolaci** [comment]
And I would like to point out that there is possibility of vetoing. So if some driver can do veto operation it must have possibility to inspect packet with zeroed FWPS_RIGHT_ACTION_WRITE flag.

I think that our driver doesn't violate any MS API documentation. And we don't have any choice - inspecting of previously blocked packets (with zeroed FWPS_RIGHT_ACTION_WRITE) is necessary for us.

And last note - because documentation is pretty unclear in this point, it's possible that WinDivert will meet some other driver with behavior similar to Avast. And result will be BSOD again.

**Comment by basil00** [comment]
Why are you bothering me with these questions?  I am not the authority on WFP behavior.  You need to raise it with Microsoft not me.

The fact of the matter is that Microsoft's WFP sample driver code appears to cause **your** driver to crash. That 's **your** problem not mine.  You also seem to be claim that the sample code is wrong.  That's a pretty big claim, but who knows, maybe you are right?  You do raise some interesting points, but I am not the person to give you an answer.  You need to immediately raise the with Microsoft and get them to release updates/fixes to their samples and/or WFP engine.

If a driver fully complies with Microsoft's recommendations and samples, and this somehow causes your driver to crash for whatever reason, then the correct remedy for the end user is pretty clear: **uninstall Avast**
